sweep/sweeper_test.go

```diff
@@ -0,0 +1,596 @@
+			Output: &wire.TxOut{
+				Value: value,
+			},
+		},
+	)
+
+	testInputCount++
+
+	return input
+}
+
+func init() {
+	// Create a set of test spendable inputs.
+	for i := 0; i < 5; i++ {
+		input := createTestInput(int64(10000 + i*500))
+
+		spendableInputs = append(spendableInputs, &input)
+	}
+}
+
+func createSweeperTestContext(t *testing.T) *sweeperTestContext {
+
+	notifier := NewMockNotifier(t)
+
+	estimator := &lnwallet.StaticFeeEstimator{
+		FeePerKW: lnwallet.SatPerKWeight(10000),
+		RelayFee: lnwallet.SatPerKWeight(1000),
+	}
+
+	publishChan := make(chan wire.MsgTx, 2)
+
+	ctx := &sweeperTestContext{
+		notifier:     notifier,
+		publishChan:  publishChan,
+		walletInputs: make(map[wire.OutPoint]*wire.MsgTx),
+		t:            t,
+		estimator:    estimator,
+		timeoutChan:  make(chan chan time.Time, 1),
+	}
+
+	var outputScriptCount byte
+	ctx.sweeper = New(&UtxoSweeperConfig{
+		GenSweepScript: func() ([]byte, error) {
+			script := []byte{outputScriptCount}
+			outputScriptCount++
+			return script, nil
+		},
+		Estimator: estimator,
+		Signer:    &mockSigner{},
+		Notifier:  notifier,
+		PublishTransaction: func(tx *wire.MsgTx) error {
+			for _, input := range tx.TxIn {
+				ctx.walletInputs[input.PreviousOutPoint] = tx
+			}
+			log.Tracef("Publishing tx %v", tx.TxHash())
+			select {
+			case publishChan <- *tx:
+			case <-time.After(defaultTestTimeout):
+				t.Fatalf("unexpected tx published")
+			}
+			return nil
+		},
+		NewBatchTimer: func() <-chan time.Time {
+			c := make(chan time.Time, 1)
+			ctx.timeoutChan <- c
+			return c
+		},
+		ChainIO: &mockChainIO{},
+		HasSpendingTx: func(op wire.OutPoint) (*wire.MsgTx, error) {
+			spendingTx, ok := ctx.walletInputs[op]
+			if !ok {
+				return nil, nil
+			}
+
+			return spendingTx, nil
+		},
+		MaxInputsPerTx: 3,
+	})
+
+	ctx.sweeper.Start()
+
+	return ctx
+}
+
+func (ctx *sweeperTestContext) tick() {
+	log.Trace("Waiting for tick to be consumed")
+	select {
+	case c := <-ctx.timeoutChan:
+		select {
+		case c <- time.Time{}:
+			log.Trace("Tick")
+		case <-time.After(defaultTestTimeout):
+			ctx.t.Fatal("tick timeout - tick not consumed")
+		}
+	case <-time.After(defaultTestTimeout):
+		ctx.t.Fatal("tick timeout - no new timer created")
+	}
+}
+
+func (ctx *sweeperTestContext) assertNoNewTimer() {
+	select {
+	case <-ctx.timeoutChan:
+		ctx.t.Fatal("no new timer expected")
+	default:
+	}
+}
+
+func (ctx *sweeperTestContext) finish() {
+	// We assume that when finish is called, sweeper has finished all its
+	// goroutines. This implies that the waitgroup is empty.
+	signalChan := make(chan struct{})
+	go func() {
+		ctx.sweeper.wg.Wait()
+		close(signalChan)
+	}()
+
+	// The only goroutine that is still expected to be running is
+	// collector(). Simulate exit of this goroutine.
+	ctx.sweeper.wg.Done()
+
+	// We now expect the Wait to succeed.
+	select {
+	case <-signalChan:
+	case <-time.After(time.Second):
+		pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
```

Comment:
very cool, have not seen this before!

Reply:
It is quite indirect, but it does the job. I use it in the nursery tests too.